### DISCUSSION TITLE
How to set position of Camera?

### DISCUSSION THREAD

**Discussion by bkeys** [openscenegraph/OpenSceneGraph#1263] [Q&A] [answered]
Hi, I bought the OpenSceneGraph 3.0 book which is very helpful; but nowhere does it tell me how to move to the camera to a specific coordinate or how to rotate it. What is the best way this can be done? I have my own custom Camera Manipulator, does it need to be done in there?
So far I have managed to manipulate my nodes by using the PositionAttitudeTransform, but I don't really see how I can do this with an osg::Camera (even though it itself is an osg::Node).

**Comment by robertosfield** [comment]
osg::Camera has a series of setVIewMatrix*() methods that allow it's view matrix to be set in various ways, you can call these in a camera manipuator or just directly in your main frame loop.  You only need to use a camera manipulator if you want to add event handling to control the camera.  The various osgGA camera manipulators will be a helpful guide.

As a general note, the OpenSceneGraph like OpenGL is now in maintenance mode, intended primarily for existing users.  For new application development I'd recommend look at the OSG's successor VullkanSceneGraph.

**Comment by bkeys** [comment]
Could you explain to me why the following code doesn't work then?
```
  float x = 100;
  float y = 1000;
  float z = 1000;
  osg::Matrixd mat;
  mat.makeTranslate(osg::Vec3(x, y, z));
  viewer.getCamera()->setViewMatrix(mat);
```
My debugger seems to say the values I set got overridden the next frame.

**Comment by XenonofArcticus** [comment]
There's a lot of things that can be controlling the camera. In a Viewer,
the Viewer's own manipulator often controls the camera.

We'd need to know a lot more about your program to answer.

Is this a simple demo you could post in its entirety?

**Reply by bkeys** [reply]
Here is my current CameraManipulator and what I am trying, I have the viewer as a global variable. The current project has several libraries tied in, if really needed I can make a minimal demo, but hopefully you can read this and see what I'm doing wrong:
Here is `Player.cpp`
```
#include "Player.hpp"
#include <osg/PositionAttitudeTransform>
#include <algorithm>
#include "Global.hpp"

Player::Player() : camera_speed(2.f),
player_camera_manip(new PlayerCameraManipulator),
pat(new osg::PositionAttitudeTransform()) {
  viewer.setCameraManipulator(player_camera_manip);
}

void PlayerCameraManipulator::translate(osg::Vec3 _dir) {
  moveUp(static_cast<double>(_dir.y()));
  moveForward(static_cast<double>(_dir.z()));
  moveRight(static_cast<double>(_dir.x()));
}
/*
bool PlayerCameraManipulator::handle(osgGA::Event *event, osg::Object *object,
                                     osg::NodeVisitor *nv) {}
bool PlayerCameraManipulator::handle(const osgGA::GUIEventAdapter &ea,
                                     osgGA::GUIActionAdapter &us) {}
*/
PlayerCameraManipulator::PlayerCameraManipulator() {
   float x = 100;
   float y = 1000;
   float z = 1000;
   osg::Matrixd mat;
   mat.makeTranslate(osg::Vec3(x, y, z));
   viewer.getCamera()->setViewMatrix(mat);
}

void PlayerCameraManipulator::rotate(osg::Vec2 _rot) {

  // world up vector
  osg::CoordinateFrame coordinateFrame = getCoordinateFrame(_eye);
  osg::Vec3d localUp = getUpVector(coordinateFrame);

  rotateYawPitch( _rotation, _rot.x(), _rot.y(), localUp );
}

void Player::process() {
process_input();
process_character_movement();

}

void Player::process_input() {
}

void Player::process_character_movement() {
}

```

And it's accompanying `Player.hpp`
```
#pragma once
#include <map>
#include <osg/Geode>
#include <osg/MatrixTransform>
#include <osgDB/ReadFile>
#include <osgGA/FirstPersonManipulator>
#include <osgGA/GUIActionAdapter>
#include <osgGA/GUIEventAdapter>
#include <osgGA/GUIEventHandler>
#include <osgViewer/Viewer>
#include <utility>
namespace osg {
  class PositionAttitudeTransform;
}


class PlayerCameraManipulator : public osgGA::FirstPersonManipulator {
public:
  PlayerCameraManipulator();
  virtual ~PlayerCameraManipulator() {
  }
  //bool handle(osgGA::Event *event, osg::Object *object, osg::NodeVisitor *nv);
  //bool handle(const osgGA::GUIEventAdapter &ea, osgGA::GUIActionAdapter &us);
  void translate(osg::Vec3 _dir);
  void rotate(osg::Vec2 _rot);
};

class Player {
public:
  Player();
  void process();
private:
  float camera_speed;
  void process_character_movement();
  void process_input();
  osg::ref_ptr<PlayerCameraManipulator> player_camera_manip;
  std::vector<std::string> _inputs;
  osg::PositionAttitudeTransform* pat;
};

```
~~I tried to override the handle functions, but it causes the program to crash for some reason I figured that was the most likely area for my transform to be overridden.~~
I can override both handle functions, and the values are still being overridden.

**Comment by bkeys** [comment, marked as answer]
So I learned the values were being overridden because of my use of a CameraManipulator; I solved it by not using a manipulator then I was able to set the values to whatever I wanted. I am using SDL2 for event handling so not using a manipulator poses little issue.
